[MAINTENANCE] Coveralls for unit-tests

Kilo59 opened the pull request:
Changes proposed in this pull request:
- publish unit-test coverage report to coveralls.io
- https://coveralls.io/github/great-expectations/great_expectations

## TODO

- in Azure
  - [x] - add `COVERALLS_REPO_TOKEN` variable to the Azure pipeline
  - [x] - pull token from the variable group
  - [ ] - don't block merge if coveralls.io is down
- in Coveralls.io
  - [x] - set `COVERAGE THRESHOLD FOR FAILURE` to `40%`
  - [x] - set `COVERAGE DECREASE THRESHOLD FOR FAILURE` to `0.5%`
  - [x] setup PR gating on the above threshold failures

<img width="712" alt="image" src="https://user-images.githubusercontent.com/13108583/187055059-5e6d93a2-284c-4efd-936b-f7b4581f9b21.png">

Kilo59 committed:
publish coveralls

`common_secrets`

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5856&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5856&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/5856/7aae00eb/995a0c546ab64fda4f94903b2e495c95ff772bb4.svg)](https://app.codesee.io/r/reviews?pr=5856&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Kilo59 commented on azure-pipelines-dependency-graph-testing.yml:
```diff
@@ -245,13 +245,23 @@ stages:
+            env:
+              CIRCLECI: 1
+              CIRCLE_BUILD_NUM: $(Build.BuildNumber)
+              COVERALLS_REPO_TOKEN: $(coveralls_repo_token)
```
We are mimicking circleci here
https://coveralls-python.readthedocs.io/en/latest/usage/configuration.html#azure-pipelines-support
https://github.com/TheKevJames/coveralls-python/issues/245

>As per [#245](https://github.com/TheKevJames/coveralls-python/issues/245), our users suggest leaving “keep this value secret” unchecked – this may be secure enough as-is, in that a user making a PR cannot access this variable.

Someone with maintainer access to `great_expectations` needs to grab the `repo_token` from the settings page so that we can add it to our azure pipeline.
https://coveralls.io/github/great-expectations/great_expectations

![Screen Shot 2022-08-29 at 9 30 34 AM](https://user-images.githubusercontent.com/13108583/187213603-848889c4-b7b9-45e6-819d-7f5d9f145f97.png)

Kilo59 committed:
install from `requirements-dev-test.txt`

Kilo59 committed:
set env `COVERALLS_REPO_TOKEN` from the `common_secrets` variable group

Kilo59 commented on azure-pipelines-dependency-graph-testing.yml:
```diff
@@ -245,13 +245,23 @@ stages:
+            env:
+              CIRCLECI: 1
+              CIRCLE_BUILD_NUM: $(Build.BuildNumber)
+              COVERALLS_REPO_TOKEN: $(common_secrets.coveralls)
```
The `coveralls` packages takes care of pulling the `COVERALLS_REPO_TOKEN` from the environment.
Need to verify that this is the correct way to pull values from variable groups.

Also, it's possible this is correct but that it being secret is causing issues? See the note below.

>As per [#245](https://github.com/TheKevJames/coveralls-python/issues/245), our users suggest leaving “keep this value secret” unchecked – this may be secure enough as-is, in that a user making a PR cannot access this variable.

## Note
The current coveralls api 422  error is the same error we get without passing any token.

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 committed:
declare variable group to use and use it

https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cshell#reference-secret-variables-in-variable-groups

Kilo59 committed:
comment out GE_pytest_opts

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 commented on azure-pipelines-dependency-graph-testing.yml:
```diff
@@ -245,13 +246,23 @@ stages:
+            env:
+              CIRCLECI: 1
+              CIRCLE_BUILD_NUM: $(Build.BuildNumber)
+              COVERALLS_REPO_TOKEN: $(coveralls)
```
This may not be an issue on this PR, but for subsequent PRs ... our unit-test coverage numbers appear to have already dropped below the `40%` threshold since there was nothing to enforce it.

https://coveralls.io/builds/52134850

So (assuming coveralls starts blocking on cov < 40%) we will need to increase our unit-testing coverage or decrease the threshold on coveralls.

Kilo59 committed:
add back `GE_pytest_opts`

Kilo59 commented on azure-pipelines-dependency-graph-testing.yml:
```diff
@@ -180,7 +180,9 @@ stages:
-          GE_pytest_opts: '--no-sqlalchemy'
+          - group: 'common_secrets'
+          - name: GE_pytest_opts
+            value: '--no-sqlalchemy'
```
Initially, I commented this `GE_pytest_opts` key in order to get the `coveralls` step working ([which it now does](https://coveralls.io/builds/52134850) 🎊 ), but I assume we need to add this back.

I think this is equivalent but LMK if that's not correct.

Kilo59 committed:
Merge branch 'develop' into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

github-actions[bot] committed:
Merge develop into coveralls

billdirks committed:
Merge branch 'develop' into coveralls

billdirks committed:
Merge branch 'develop' into coveralls

Kilo59 committed:
fix conflicts

Kilo59 committed:
remove extra dep install step

Kilo59 committed:
more descriptive step name

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 committed:
Merge remote-tracking branch 'origin/coveralls' into coveralls

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 committed:
Merge branch 'develop' into coveralls

Kilo59 commented on azure-pipelines-dev.yml:
```diff
@@ -243,13 +245,23 @@ stages:
+              pip install coveralls
+              export CIRCLE_BRANCH=$BUILD_SOURCEBRANCH
+              coveralls
+            displayName: 'coveralls for Unit Tests'
+            env:
+              CIRCLECI: 1
+              CIRCLE_BUILD_NUM: $(Build.BuildNumber)
+              COVERALLS_REPO_TOKEN: $(coveralls)
```
Is this the correct way to prevent this step from breaking the pipeline?
Like if `coveralls.io` goes down.

```suggestion
            displayName: 'coveralls for Unit Tests'
            continueOnError: true 
            env:
              CIRCLECI: 1
              CIRCLE_BUILD_NUM: $(Build.BuildNumber)
              COVERALLS_REPO_TOKEN: $(coveralls)
```

Kilo59 committed:
Merge branch 'develop' into coveralls
